flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreExpireImpl.java

```diff
@@ -101,18 +120,28 @@ public void expire() {
         }
 
         // no snapshot can be retained, expire all but last one
-        expireUntil(latestSnapshotId);
+        expireUntil(earliest, latestSnapshotId);
     }
 
-    private void expireUntil(long endExclusiveId) {
-        if (endExclusiveId <= Snapshot.FIRST_SNAPSHOT_ID) {
+    private void expireUntil(long earliestId, long endExclusiveId) {
+        if (endExclusiveId <= earliestId) {
+            // write hint file if not exists
+            Path hint = new Path(pathFactory.snapshotDirectory(), SnapshotFinder.EARLIEST);
+            try {
+                if (!hint.getFileSystem().exists(hint)) {
```

Comment:
Do we need to check this? We're creating and renaming new files in `commitEarliestHint` so it doesn't matter if the old file exists.

Reply:
If it happens to be after the deletion before the renaming, then a job comes to read the data, it will see that there is no hint file

Reply:
No expire happens:
write the hint file in order to see the earliest snapshot directly next time
should avoid duplicate writes when the file exists

Reply:
Yes, but this will have no impact on the file content and in the meantime make the logic here more simple.

Reply:
I think the logic is not complicated, which avoids N earliest snapshots written before the appearance of expiration.